lib/babe/verify.go

```diff
@@ -327,62 +327,58 @@ func (b *verifier) verifyAuthorshipRight(header *types.Header) error {
 		return ErrBadSignature
 	}
 
-	// check if the producer has equivocated, ie. have they produced a conflicting block?
-	// hashes is hashes of all blocks with same block number as header.Number
-	hashes := b.blockState.GetAllBlocksAtDepth(header.ParentHash)
+	equivocated, err := b.verifyBlockEquivocation(header)
+	if err != nil {
+		return fmt.Errorf("could not verify block equivocation for header %s: %w", header.Hash(), err)
+	}
+	if equivocated {
+		return ErrProducerEquivocated
+	}
 
-	for _, currentHash := range hashes {
-		currentHeader, err := b.blockState.GetHeader(currentHash)
-		if err != nil {
-			return fmt.Errorf("failed get header %s", err)
-		}
+	return nil
+}
 
-		currentBlockProducerIndex, err := getAuthorityIndex(currentHeader)
-		if err != nil {
-			return fmt.Errorf("failed to get authority index %s", err)
-		}
+// verifyBlockEquivocation checks if given block's author has occupied corresponding slot more than once.
+// It returns true if block was equivocated.
+func (b *verifier) verifyBlockEquivocation(header *types.Header) (bool, error) {
```

Comment:
nit I'd suggest naming your return so there is no need to mention it returns true if the the block is equivocated:

```suggestion
func (b *verifier) verifyBlockEquivocation(header *types.Header) (equivocated bool, err error) {
```

Reply:
I prefer to explicitly declare variables rather than declaring with named returns.

Reply:
I'd still argue that documentation with code is better than documentation with comments

> Good code is its own best documentation. As you’re about to add a comment, ask yourself, 'How can I improve the code so that this comment isn’t needed?' Improve the code and then document it to make it even clearer

Reply:
Naming return variables in the signature leads to two things
- you can get away with not explicitly declaring variables
- you can get away with having naked returns instead of explicit returns.

I can't really keep on arguing about every small small things. Even if it is a nit comment, I have to resolve it. I am very humbly requesting you to stop giving out reviews based on preferences and opinions.